server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1596,6 +1605,32 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
         }
     }
 
+    /**
+     * This method tries to rewrite a query without using a searcher (before we get a valid one). It takes advantage of the fact that
+     * we can skip some shards in the query phase because we have enough information in the index mapping to decide the 'can match'
+     * outcome. One such example are queries like {@link org.elasticsearch.index.query.MatchPhraseQueryBuilder} queries which are rewritten
+     * to {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} after being rewritten to
+     * {@link org.elasticsearch.index.query.TermQueryBuilder}. Other queries have a similar pattern, especially when using
+     * {@link org.elasticsearch.index.mapper.ConstantFieldType} or unmapped fields, they all rewrite to
+     * {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} (if conditions are met). This allows us to avoid extra work other than
```

Comment:
Maybe this text of this example can be simplified? Something like: `On such example is a term based query against a constant keyword field. This queries can rewrite themselves to a `MatchNoneQueryBuilder`.